File: IdentityServer/Startup.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.HttpsPolicy;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Storage;
using Contexts;
using DatabaseTypes;
using Models.Authentication;
using Microsoft.AspNetCore.Authentication;
using IdentityServer4.Stores;
using IdentityServer4.Services;
using Microsoft.AspNetCore.HttpOverrides;
using Prometheus;
using System.Net;

namespace IdentityServer
{
    public class Startup
    {
        public Startup(IConfiguration configuration)
        {
            Configuration = configuration;
        }

        public IConfiguration Configuration { get; }

        // This method gets called by the runtime. Use this method to add services to the container.
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddRazorPages();
            services.AddMvc();

            services.AddHealthChecks()
                .ForwardToPrometheus();

            services.Configure<CouchbaseConfig>(this.Configuration.GetSection("Couchbase"))
                    .AddSingleton<CouchbaseCluster>()
                    .AddSingleton<CouchbaseBuckets>()
                    .AddSingleton<IUserContext, UserContext>()
                    .AddSingleton<IContext<UserGroups>, UserGroupsContext>()
                    .AddSingleton<IContext<UserItems>, UserItemsContext>()
                    .AddSingleton<IContext<Trash>, TrashContext>()
                    .AddSingleton<IUserRoleContext, UserRoleContext>();

            services.AddOidcStateDataFormatterCache();

            services.AddIdentity<UserIdentity, RoleIdentity>()
                .AddRoleStore<RoleStore>()
                .AddUserStore<UserStore>();

            services.AddSingleton<IClientStore, ClientStore>();

            services.AddSingleton<ICorsPolicyService, CorsPolicyService>();

            services.AddIdentityServer()
                    .AddDeveloperSigningCredential()
                    .AddInMemoryApiResources(Config.ApiResources)
                    .AddInMemoryIdentityResources(Config.IdentityResources)
                    .AddAspNetIdentity<UserIdentity>();
            services.AddAuthentication()
                    .AddIdentityServerJwt();

        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }
            else
            {
                app.UseExceptionHandler("/Error");
                // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
                app.UseHsts();

                ForwardedHeadersOptions forwardedHeadersOptions = new ForwardedHeadersOptions
                {
                    ForwardedHeaders = ForwardedHeaders.XForwardedFor | ForwardedHeaders.XForwardedProto,
                    KnownProxies = { IPAddress.Parse("10.5.0.7") }
                };

                app.UseForwardedHeaders(forwardedHeadersOptions);
            }

            app.UseHttpsRedirection();

            app.UseStaticFiles();
            app.UseRouting();
            app.UseHttpMetrics();
            app.UseAuthentication();
            app.UseAuthorization();
            app.UseIdentityServer();

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapDefaultControllerRoute();
                endpoints.MapMetrics();
                endpoints.MapHealthChecks("/health");
            });
        }
    }
}

File: CSharpGenerator/Startup.fs
namespace CSharpGenerator

open Microsoft.AspNetCore.Builder
open Microsoft.AspNetCore.Hosting
open Microsoft.Extensions.Configuration
open Microsoft.Extensions.DependencyInjection
open Microsoft.Extensions.Hosting
open System.Text.Json.Serialization
open Microsoft.Extensions.Caching.Memory
open Prometheus


type Startup private () =
    new (configuration: IConfiguration) as this =
        Startup() then
        this.Configuration <- configuration

    // This method gets called by the runtime. Use this method to add services to the container.
    member this.ConfigureServices(services: IServiceCollection) =
        // Add framework services.
        services.AddControllers()
        |> fun x -> x.AddJsonOptions(fun options ->
                options.JsonSerializerOptions.Converters.Add(JsonFSharpConverter()))
        |> ignore

        services.AddHealthChecks()
        |> fun x -> x.ForwardToPrometheus()
        |> ignore

        services
        |> fun x -> x.AddSingleton<Processor>()
        |> fun x -> x.AddSingleton<CSharpGenerator>()
        |> fun x -> x.AddSingleton<IMemoryCache, MemoryCache>()
        |> fun x -> x.AddSingleton<ISimpleValidator<IntValidatorOptions>, IntValidator>()
        |> fun x -> x.AddSingleton<ISimpleValidator<FloatValidatorOptions>, FloatValidator>()
        |> fun x -> x.AddSingleton<ISimpleValidator<StringValidatorOptions>, StringValidator>()
        |> ignore

    // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
    member this.Configure(app: IApplicationBuilder, env: IWebHostEnvironment) =
        if (env.IsDevelopment()) then
            app.UseDeveloperExceptionPage() |> ignore

        app.UseHttpsRedirection() |> ignore
        app.UseRouting() |> ignore
        app.UseHttpMetrics() |> ignore

        app.UseAuthorization() |> ignore

        app.UseEndpoints(fun endpoints ->
            endpoints.MapControllers() |> ignore
            endpoints.MapMetrics() |> ignore
            endpoints.MapHealthChecks("/health") |> ignore
            ) |> ignore

    member val Configuration : IConfiguration = null with get, set
